### PR TITLE
RUBY-999: Use constant time hash comparison

### PR DESCRIPTION
**Pull request by durran**

### CODE REVIEW

File: lib/mongo/auth/scram/conversation.rb [hunk 1]
```diff
@@ -429,9 +429,15 @@ def xor(first, second)
           first.bytes.zip(second.bytes).map{ |(a,b)| (a ^ b).chr }.join('')
         end
 
+        def secure_compare(a, b)
```
Comment (behackett) [inline lib/mongo/auth/scram/conversation.rb]: Using the word "secure" is too strong of a claim (the original attempt to implement this for cpython made the same claim and was rejected). compare_digest seems better.

File: lib/mongo/auth/scram/conversation.rb [hunk 2]
```diff
@@ -429,9 +429,15 @@ def xor(first, second)
...
+          check = a.bytesize ^ b.bytesize
+          a.bytes.zip(b.bytes){ |x, y| check |= x ^ y.to_i }
```
Comment (behackett) [inline lib/mongo/auth/scram/conversation.rb]: What does zip do if the string lengths aren't equal? MIght be better to take the hash from the server and compare it to itself in that case.

Comment (durran) [inline lib/mongo/auth/scram/conversation.rb]: If they lengths are not equal then null bytes are used to pad the shorter to the length of the longer.

File: lib/mongo/auth/scram/conversation.rb (hunk 1 above)
Comment (durran) [inline lib/mongo/auth/scram/conversation.rb]: Renamed.

**Comment by behackett** [comment]
LGTM

**Pull request merged**